e2e/e2e_openssl_test.go

```diff
@@ -0,0 +1,245 @@
+	// #nosec G204
+	cmd := exec.CommandContext(c.ctx, "openssl", args...)
+	var inner net.Conn
+	inner, c.clientConn = net.Pipe()
+	cmd.Stdin = inner
+	cmd.Stdout = inner
+	cmd.Stderr = os.Stderr
+	if err := cmd.Start(); err != nil {
+		c.errChan <- err
+		_ = inner.Close()
+		return
+	}
+
+	simpleReadWrite(c.errChan, c.clientChan, c.clientConn, c.messageRecvCount)
+}
+
+func ciphersOpenSSL(cfg *dtls.Config) string {
+	// See https://tls.mbed.org/supported-ssl-ciphersuites
+	translate := map[dtls.CipherSuiteID]string{
+		dtls.TLS_ECDHE_ECDSA_WITH_AES_128_CCM:        "ECDHE-ECDSA-AES128-CCM",
+		dtls.TLS_ECDHE_ECDSA_WITH_AES_128_CCM_8:      "ECDHE-ECDSA-AES128-CCM8",
+		dtls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256: "ECDHE-ECDSA-AES128-GCM-SHA256",
+		dtls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256:   "ECDHE-RSA-AES128-GCM-SHA256",
+
+		dtls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA: "ECDHE-ECDSA-AES256-SHA",
+		dtls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA:   "ECDHE-RSA-AES128-SHA",
+
+		dtls.TLS_PSK_WITH_AES_128_CCM:        "PSK-AES128-CCM",
+		dtls.TLS_PSK_WITH_AES_128_CCM_8:      "PSK-AES128-CCM8",
+		dtls.TLS_PSK_WITH_AES_128_GCM_SHA256: "PSK-AES128-GCM-SHA256",
+	}
+
+	var ciphers []string
+	for _, c := range cfg.CipherSuites {
+		if text, ok := translate[c]; ok {
+			ciphers = append(ciphers, text)
+		}
+	}
+	return strings.Join(ciphers, ";")
+}
+
+func writeTempPEM(cfg *dtls.Config) (string, string, error) {
+	certOut, err := ioutil.TempFile("", "cert.pem")
+	if err != nil {
+		return "", "", fmt.Errorf("failed to create temporary file: %w", err)
+	}
+	keyOut, err := ioutil.TempFile("", "key.pem")
+	if err != nil {
+		return "", "", fmt.Errorf("failed to create temporary file: %w", err)
+	}
+
+	cert := cfg.Certificates[0]
+	derBytes := cert.Certificate[0]
+	if err = pem.Encode(certOut, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes}); err != nil {
+		return "", "", fmt.Errorf("failed to write data to cert.pem: %w", err)
+	}
+	if err = certOut.Close(); err != nil {
+		return "", "", fmt.Errorf("error closing cert.pem: %w", err)
+	}
+
+	priv := cert.PrivateKey
+	var privBytes []byte
+	privBytes, err = x509.MarshalPKCS8PrivateKey(priv)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to marshal private key: %w", err)
+	}
+	if err = pem.Encode(keyOut, &pem.Block{Type: "PRIVATE KEY", Bytes: privBytes}); err != nil {
+		return "", "", fmt.Errorf("failed to write data to key.pem: %w", err)
+	}
+	if err = keyOut.Close(); err != nil {
+		return "", "", fmt.Errorf("error closing key.pem: %w", err)
+	}
+	return certOut.Name(), keyOut.Name(), nil
+}
+
+func TestPionOpenSSLE2ESimple(t *testing.T) {
+	t.Run("OpenSSLServer", func(t *testing.T) {
+		testPionE2ESimple(t, serverOpenSSL, clientPion)
+	})
+	t.Run("OpenSSLClient", func(t *testing.T) {
+		testPionE2ESimple(t, serverPion, clientOpenSSL)
+	})
+}
+func TestPionOpenSSLE2ESimplePSK(t *testing.T) {
+	// TODO: make PSK work with openssl server
+	/*
+		t.Run("OpenSSLServer", func(t *testing.T) {
```

Comment:
This seems already working on my environment.
```
$ go test ./... -run TestPionOpenSSLE2ESimplePSK -tags openssl -v -count 1
=== RUN   TestPionOpenSSLE2ESimplePSK
=== RUN   TestPionOpenSSLE2ESimplePSK/OpenSSLServer
=== RUN   TestPionOpenSSLE2ESimplePSK/OpenSSLServer/TLS_PSK_WITH_AES_128_CCM
139968244041536:error:141023E8:SSL routines:dtls1_read_bytes:reason(1000):ssl/record/rec_layer_d1.c:611:SSL alert number 0
=== RUN   TestPionOpenSSLE2ESimplePSK/OpenSSLServer/TLS_PSK_WITH_AES_128_CCM_8
140499049539392:error:141023E8:SSL routines:dtls1_read_bytes:reason(1000):ssl/record/rec_layer_d1.c:611:SSL alert number 0
=== RUN   TestPionOpenSSLE2ESimplePSK/OpenSSLServer/TLS_PSK_WITH_AES_128_GCM_SHA256
140184440280896:error:141023E8:SSL routines:dtls1_read_bytes:reason(1000):ssl/record/rec_layer_d1.c:611:SSL alert number 0
=== RUN   TestPionOpenSSLE2ESimplePSK/OpenSSLClient
=== RUN   TestPionOpenSSLE2ESimplePSK/OpenSSLClient/TLS_PSK_WITH_AES_128_CCM
Can't use SSL_get_servername
140229900810048:error:141023E8:SSL routines:dtls1_read_bytes:reason(1000):ssl/record/rec_layer_d1.c:611:SSL alert number 0
=== RUN   TestPionOpenSSLE2ESimplePSK/OpenSSLClient/TLS_PSK_WITH_AES_128_CCM_8
Can't use SSL_get_servername
=== RUN   TestPionOpenSSLE2ESimplePSK/OpenSSLClient/TLS_PSK_WITH_AES_128_GCM_SHA256
Can't use SSL_get_servername
140696154187584:error:141023E8:SSL routines:dtls1_read_bytes:reason(1000):ssl/record/rec_layer_d1.c:611:SSL alert number 0
--- PASS: TestPionOpenSSLE2ESimplePSK (3.67s)
    --- PASS: TestPionOpenSSLE2ESimplePSK/OpenSSLServer (2.52s)
        --- PASS: TestPionOpenSSLE2ESimplePSK/OpenSSLServer/TLS_PSK_WITH_AES_128_CCM (0.71s)
        --- PASS: TestPionOpenSSLE2ESimplePSK/OpenSSLServer/TLS_PSK_WITH_AES_128_CCM_8 (0.71s)
        --- PASS: TestPionOpenSSLE2ESimplePSK/OpenSSLServer/TLS_PSK_WITH_AES_128_GCM_SHA256 (0.71s)
    --- PASS: TestPionOpenSSLE2ESimplePSK/OpenSSLClient (1.14s)
        --- PASS: TestPionOpenSSLE2ESimplePSK/OpenSSLClient/TLS_PSK_WITH_AES_128_CCM (0.24s)
        --- PASS: TestPionOpenSSLE2ESimplePSK/OpenSSLClient/TLS_PSK_WITH_AES_128_CCM_8 (0.25s)
        --- PASS: TestPionOpenSSLE2ESimplePSK/OpenSSLClient/TLS_PSK_WITH_AES_128_GCM_SHA256 (0.25s)
PASS
ok  	github.com/pion/dtls/v2/e2e	3.682s
```

Reply:
I believe I'm seeing a bunch of errors in the output? All those `error` lines from OpenSSL?

Reply:
Yes, errors are from openssl and `Can't use SSL_get_servername` means that SNI is not supported.

Reply:
What did you mean by tests are working in that case? Is it that they run, or that they're correct and pass?

Seeing stuff like `139968244041536:error:141023E8:SSL routines:dtls1_read_bytes:reason(1000):ssl/record/rec_layer_d1.c:611:SSL alert number 0` but then a `PASS` seems to suggest something is wrong. If that error message is expected, then we should probably make that clear and not have it show up.

Reply:
```
140696154187584:error:141023E8:SSL routines:dtls1_read_bytes:reason(1000):ssl/record/rec_layer_d1.c:611:SSL alert number 0
```
is close_notify.

BTW, pion/dtls sends `close_notify` as fatal level, but OpenSSL sends `close_notify` as warning level. TLS 1.2 doesn't explicitly specify the level of `close_notify`. It was specified as warning level in TLS 1.0 RFC.

Reply:
I opened an issue about this. https://github.com/pion/dtls/issues/195

Reply:
FYI, all other OpenSSL E2E tests also shows the same errors.

Reply:
From the functional prespective tests are passing. Now we should check all these OpenSSL error messages to know if their reason is some bad implementation on pion/dtls or they can be ignored.